libs/contrib/Data/HVect.idr

```diff
@@ -0,0 +1,134 @@
+deleteAt FZ (x :: xs) = xs
+deleteAt (FS FZ) (x :: (y :: xs)) = x :: xs
+deleteAt (FS (FS j)) (x :: (y :: xs)) = x :: deleteAt (FS j) (y :: xs)
+
+||| Replace an element in an HVect.
+|||
+||| ```idris example
+||| > replaceAt 0 "firstString" (the (HVect _) [1, "string"])
+||| ["firstString", "string"]
+||| ```
+public export
+replaceAt : (1 i : Fin k) -> t -> HVect ts -> HVect (replaceAt i t ts)
+replaceAt FZ y (x :: xs) = y :: xs
+replaceAt (FS j) y (x :: xs) = x :: replaceAt j y xs
+
+||| Update an element in an HVect.
+|||
+||| ```idris example
+||| > updateAt 0 (const True) (the (HVect _) [1, "string"])
+||| [True, "string"]
+||| ```
+public export
+updateAt : (1 i : Fin k) -> (index i ts -> t) -> HVect ts -> HVect (replaceAt i t ts)
+updateAt FZ f (x :: xs) = f x :: xs
+updateAt (FS j) f (x :: xs) = x :: updateAt j f xs
+
+||| Append two `HVect`s.
+|||
+||| ```idris example
+||| > (the (HVect _) [1]) ++ (the (HVect _) ["string"])
+||| [1, "string"]
+||| ```
+public export
+(++) : HVect ts -> HVect us -> HVect (ts ++ us)
+(++) [] ys = ys
+(++) (x :: xs) ys = x :: (xs ++ ys)
+
+public export
+Eq (HVect []) where
+  [] == [] = True
+
+public export
+(Eq t, Eq (HVect ts)) => Eq (HVect (t :: ts)) where
+  (x :: xs) == (y :: ys) = x == y && xs == ys
+
+public export
+consInjective1 : {xs, ys: HVect ts} -> {x, y: a} -> x :: xs = y :: ys -> x = y
+consInjective1 Refl = Refl
+
+public export
+consInjective2 : {xs, ys: HVect ts} -> {x, y: a} -> x :: xs = y :: ys -> xs = ys
+consInjective2 Refl = Refl
+
+public export
+DecEq (HVect []) where
+  decEq [] [] = Yes Refl
+
+public export
+(DecEq t, DecEq (HVect ts)) => DecEq (HVect (t :: ts)) where
+  decEq (x :: xs) (y :: ys) with (decEq x y)
+    decEq (z :: xs) (z :: ys) | Yes Refl with (decEq xs ys)
+      decEq (z :: zs) (z :: zs) | Yes Refl | Yes Refl = Yes Refl
+      decEq (z :: xs) (z :: ys) | Yes Refl | No contra = No (contra . consInjective2)
+    decEq (x :: xs) (y :: ys) | No contra = No (contra . consInjective1)
+
+public export
+interface Shows (k : Nat) (ts : Vect k Type) where
+  shows : HVect ts -> Vect k String
+
+public export
+Shows Z [] where
+  shows [] = []
+
+public export
+(Show t, Shows len ts) => Shows (S len) (t :: ts) where
+  shows (x :: xs) = show x :: shows xs
+
+public export
+(Shows len ts) => Show (HVect ts) where
+  show xs = "[" ++ (pack . intercalate [','] . map unpack . toList $ shows xs) ++ "]"
+
+||| Extract an arbitrary element of the correct type.
+||| @ t the goal type
+public export
+get : HVect ts -> {auto 1 p : Elem t ts} -> t
+get (x :: xs) {p = Here} = x
+get (x :: xs) {p = (There p')} = get xs
+
+||| Replace an element with the correct type.
+public export
+put : t -> HVect ts -> {auto 1 p : Elem t ts} -> HVect ts
+put y (x :: xs) {p = Here} = y :: xs
+put y (x :: xs) {p = (There p')} = x :: put y xs
+
+||| Update an element with the correct type.
+public export
+update : (t -> u) -> HVect ts -> {auto 1 p : Elem t ts} -> HVect (replaceByElem ts p u)
```

Comment:
I find it strange that `put` is homogeneous but `update` is not.

More generally, it seems a bit strange to me to have both:
* a `Fin`-based interface (that justifies using `Vect` as the index for `HVect`)
* an `Elem`-based interface (that would be well-suited for a `List`-based `HVect` aka `All`)

Reply:
We could have homogeneous and heterogeneous for both (but... naming is hard).

`Elem` as an implicit proof is very handy for heterogeneous lists, disregarding their implementation, as it allows you to refer to a specific position for a given type.

I also use the `Fin` based access in my codebase, I don't remember exactly when (and more importantly why) though.

Reply:
Do you think implementing homogeneous & heterogeneous versions of put and update in this would be beneficial, or is the current implementation of put & update good?